File: docker-gradle-plugin-main/src/main/java/com/github/godfather1103/gradle/entity/Git.java
package com.github.godfather1103.gradle.entity;

import com.spotify.docker.client.exceptions.DockerException;
import org.eclipse.jgit.api.Status;
import org.eclipse.jgit.api.errors.GitAPIException;
import org.eclipse.jgit.lib.ObjectId;
import org.eclipse.jgit.lib.Ref;
import org.eclipse.jgit.lib.Repository;
import org.eclipse.jgit.storage.file.FileRepositoryBuilder;
import org.gradle.api.GradleException;

import java.io.IOException;

import static com.google.common.base.Strings.isNullOrEmpty;

/**
 * <p>Title:        Godfather1103's Github</p>
 * <p>Copyright:    Copyright (c) 2022</p>
 * <p>Company:      https://github.com/godfather1103</p>
 *
 * @author 作者: Jack Chu E-mail: chuchuanbao@gmail.com<br/>
 * <p>
 * 创建时间：2022/12/13 17:14
 * @version 1.0
 * @since 1.0
 */
public class Git {

    private Repository repo;

    public Git() throws IOException {
        final FileRepositoryBuilder builder = new FileRepositoryBuilder();
        // scan environment GIT_* variables
        builder.readEnvironment();
        // scan up the file system tree
        builder.findGitDir();
        // if getGitDir is null, then we are not in a git repository
        repo = builder.getGitDir() == null ? null : builder.build();
    }

    public boolean isRepository() {
        return repo != null;
    }

    public Repository getRepo() {
        return repo;
    }

    void setRepo(final Repository repo) {
        this.repo = repo;
    }

    public String getCommitId()
            throws GitAPIException, DockerException, IOException, GradleException {

        if (repo == null) {
            throw new GradleException(
                    "Cannot tag with git commit ID because directory not a git repo");
        }

        final StringBuilder result = new StringBuilder();

        try {
            // get the first 7 characters of the latest commit
            final ObjectId head = repo.resolve("HEAD");
            if (head == null || isNullOrEmpty(head.getName())) {
                return null;
            }

            result.append(head.getName().substring(0, 7));
            final org.eclipse.jgit.api.Git git = new org.eclipse.jgit.api.Git(repo);

            // append first git tag we find
            for (final Ref gitTag : git.tagList().call()) {
                if (gitTag.getObjectId().equals(head)) {
                    // name is refs/tag/name, so get substring after last slash
                    final String name = gitTag.getName();
                    result.append(".");
                    result.append(name.substring(name.lastIndexOf('/') + 1));
                    break;
                }
            }

            // append '.DIRTY' if any files have been modified
            final Status status = git.status().call();
            if (status.hasUncommittedChanges()) {
                result.append(".DIRTY");
            }
        } finally {
            repo.close();
        }

        return result.length() == 0 ? null : result.toString();
    }

}

File: docker-gradle-plugin-main/src/main/kotlin/com/github/godfather1103/gradle/entity/Git.kt
package com.github.godfather1103.gradle.entity

import com.github.dockerjava.api.exception.DockerException
import org.eclipse.jgit.api.errors.GitAPIException
import org.eclipse.jgit.lib.Repository
import org.eclipse.jgit.storage.file.FileRepositoryBuilder
import org.gradle.api.GradleException
import java.io.IOException
import java.util.*

class Git {

    var repo: Optional<Repository> = Optional.empty()

    constructor() {
        val builder = FileRepositoryBuilder()
        builder.readEnvironment()
        builder.findGitDir()
        if (builder.gitDir != null) {
            repo = Optional.ofNullable(builder.build())
        }
    }

    fun isRepository(): Boolean {
        return repo.isPresent
    }

    @Throws(GitAPIException::class, DockerException::class, IOException::class, GradleException::class)
    fun getCommitId(): String {
        if (!repo.isPresent) {
            throw GradleException(
                "Cannot tag with git commit ID because directory not a git repo"
            )
        }
        val result = StringBuilder()
        repo.get().use {
            val head = Optional.ofNullable(it.resolve("HEAD"))
                .map { i -> i.name }
                .orElse("")
            if (head.isEmpty()) {
                return ""
            }
            result.append(head.substring(0, 8))
            val git = org.eclipse.jgit.api.Git(it)
            // append first git tag we find
            for (gitTag in git.tagList().call()) {
                if (gitTag.objectId.equals(head)) {
                    // name is refs/tag/name, so get substring after last slash
                    val name = gitTag.name
                    result.append(".")
                    result.append(name.substring(name.lastIndexOf('/') + 1))
                    break
                }
            }
            // append '.DIRTY' if any files have been modified
            val status = git.status().call()
            if (status.hasUncommittedChanges()) {
                result.append(".DIRTY")
            }
        }
        return if (result.isEmpty()) "" else result.toString()
    }
}